subsys/bluetooth/services/dis.c

```diff
@@ -222,14 +354,169 @@ static int dis_set(const char *name, size_t len_rd,
 		len = read_cb(store, &dis_sw_rev, sizeof(dis_sw_rev) - 1);
 		if (len < 0) {
 			LOG_ERR("Failed to read software revision from storage"
-				       " (err %zd)", len);
+				" (err %zd)",
+				len);
 		} else {
 			dis_sw_rev[len] = '\0';
 
 			LOG_DBG("Software revision set to %s", dis_sw_rev);
 		}
 		return 0;
 	}
+#endif
+#if defined(CONFIG_BT_DIS_UDI)
+	const size_t currently_used_udi_merged_space =
+		1 + BT_DIS_STRLEN(BT_DIS_UDI_LABEL_STR_REF) + BT_DIS_STRLEN(BT_DIS_UDI_DI_STR_REF) +
+		BT_DIS_STRLEN(BT_DIS_UDI_ISSUER_STR_REF) +
+		BT_DIS_STRLEN(BT_DIS_UDI_AUTHORITY_STR_REF) + BT_DIS_UDI_FLAG_LABEL +
+		BT_DIS_UDI_FLAG_DI + BT_DIS_UDI_FLAG_ISSUER + BT_DIS_UDI_FLAG_AUTHORITY;
+
+	if (!strncmp(name, "udi_label", nlen)) {
+		size_t in_use_without_this = currently_used_udi_merged_space -
+					     BT_DIS_STRLEN(BT_DIS_UDI_LABEL_STR_REF) -
+					     BT_DIS_UDI_FLAG_LABEL;
+		if (BT_ATT_MAX_ATTRIBUTE_LEN < in_use_without_this + strlen(store) + 1) {
+			LOG_ERR("Failed to set UDI Label. Not enough space. The sum of the 4 DIS "
+				"UDI for "
+				"Medical Devices strings may not exceed the maximum attribute "
+				"length.");
+			return 0;
+		}
+
+		len = read_cb(store, &dis_udi_label, sizeof(dis_udi_label) - 1);
+		if (len < 0) {
+			LOG_ERR("Failed to read UDI Label from storage"
+				" (err %zd)",
+				len);
+		} else {
+			dis_udi_label[len] = '\0';
+
+			LOG_DBG("UDI Label set to %s", dis_udi_label);
+		}
+		return 0;
+	}
+
+	if (!strncmp(name, "udi_di", nlen)) {
+		size_t in_use_without_this = currently_used_udi_merged_space -
+					     BT_DIS_STRLEN(BT_DIS_UDI_DI_STR_REF) -
+					     BT_DIS_UDI_FLAG_DI;
+		if (BT_ATT_MAX_ATTRIBUTE_LEN < in_use_without_this + strlen(store) + 1) {
+			LOG_ERR("Failed to set UDI Label. Not enough space. The sum of the 4 DIS "
+				"UDI for "
+				"Medical Devices strings may not exceed the maximum attribute "
+				"length.");
+			return 0;
+		}
+
+		len = read_cb(store, &dis_udi_di, sizeof(dis_udi_di) - 1);
+		if (len < 0) {
+			LOG_ERR("Failed to read UDI Device Identifier from storage"
+				" (err %zd)",
+				len);
+		} else {
+			dis_udi_di[len] = '\0';
+
+			LOG_DBG("UDI Device Identifier set to %s", dis_udi_di);
+		}
+		return 0;
+	}
+
+	if (!strncmp(name, "udi_issuer", nlen)) {
+		size_t in_use_without_this = currently_used_udi_merged_space -
+					     BT_DIS_STRLEN(BT_DIS_UDI_ISSUER_STR_REF) -
+					     BT_DIS_UDI_FLAG_ISSUER;
+		if (BT_ATT_MAX_ATTRIBUTE_LEN < in_use_without_this + strlen(store) + 1) {
+			LOG_ERR("Failed to set UDI Label. Not enough space. The sum of the 4 DIS "
+				"UDI for "
+				"Medical Devices strings may not exceed the maximum attribute "
+				"length.");
+			return 0;
+		}
+
+		len = read_cb(store, &dis_udi_issuer, sizeof(dis_udi_issuer) - 1);
+		if (len < 0) {
+			LOG_ERR("Failed to read UDI Issuer from storage"
+				" (err %zd)",
+				len);
+		} else {
+			dis_udi_issuer[len] = '\0';
+
+			LOG_DBG("UDI Issuer set to %s", dis_udi_issuer);
+		}
+		return 0;
```

Comment:
There is a lot of repeated code for the old characteristics as well.
We could definitely move the verification to a separate method, but maybe the storage and logging should stay, so that we do not have to pass the name of the specific string to the method. Then we would waste resources processing that when we do not need to log. Although, maybe a generic error message is sufficient.

Reply:
You can pass a string like "UDI Issuer" or "System ID identifier" to a common function to get specific log statements

Reply:
Yes, but that would technically waste some resources as the normal use-case is without logging enabled.

Reply:
That's true.